# ภาษา Java

ภาษา Java เป็นภาษาสำหรับงานทั่วไปที่สนับสนุนแนวคิดการโปรแกรมเชิงวัตถุ (Object-Oriented Programmin [OOP]) พัฒนาขึ้นโดย James Gosling และทีมงานที่ Sun Microsystems และเปิดตัวในปี 1995

โครงการพัฒนาภาษาใหม่ของ Sun Microsystems มีชื่อดั้งเดิมว่า Oak ตามชื่อของต้นไม้ที่อยู่หน้าห้องทำงานของ Gosling แต่เปลี่ยนเป็น Green และเป็น Java ในที่สุด โดยตั้งตามกาแฟจากอินโดนีเซีย (กาแฟชวา [Java]) เราจะเห็นได้จากสัญลักษณ์ของภาษา Java ที่มักจะเป็นรูปถ้วยกาแฟที่กรุ่นควัน

รูปแบบของภาษา Java มีรากฐานมาจากภาษา C และ C++ แต่ออกแบบให้สามารถเขียนโปรแกรมที่ทำงานถูกต้องได้ง่ายขึ้นโดยการตัดการทำงานในระดับต่ำบางอย่างออกไป

ภาษา Java ออกแบบมาให้มีจุดเด่นที่สำคัญดังนี้

- เป็นภาษาเชิงวัตถุที่เข้าใจง่ายและใช้งานง่าย เมื่อเทียบกับภาษาอย่าง C++ ในสมัยนั้น ภาษา Java ถือว่าเรียบง่ายกว่ามาก แต่ก็มีรูปแบบภาษาที่ใกล้เคียง C และ C++ ซึ่งทำให้คนที่ใช้ภาษาเหล่านั้นอยู่ก่อนสามารถเรียนรู้ได้ง่ายขึ้น
- Write Once, Run Anywhere โปรแกรมที่เขียนขึ้นและคอมไพล์แล้วสามารถนำไปรันที่ระบบไหนก็ได้ ถือเป็นคติพจน์ของ Java ที่ได้รับการโฆษณาอย่างมากในช่วงแรก ในทางปฏิบัติก็มีข้อจำกัดเฉพาะระบบที่ทำให้การใช้งานในระบบที่ต่างกันแตกต่างกันออกไปได้บ้าง
- มีการจัดการหน่วยความจำด้วยตัวเก็บขยะ (garbage collector) ภาษาอย่าง C/C++ ผู้เขียนโปรแกรมต้องจัดการทำลายข้อมูลที่สร้างขึ้นมาหลังจากเลิกใช้แล้วด้วยตัวเอง ซึ่งเป็นจุดที่มักจะสร้างความผิดพลาดในการทำงานของโปรแกรมได้มากที่สุดจุดหนึ่ง ภาษา Java ใช้ตัวเก็บขยะจัดการทำลายข้อมูลที่ไม่ใช้แล้วโดยอัตโนมัติ ทำให้ผู้เขียนโปรแกรมไม่ต้องทำเอง ลดความเสี่ยงต่อความผิดพลาดลงได้อีกจุดหนึ่ง
- ตัวแปลภาษา ระบบสนับสนุน เครื่องมือต่าง ๆ ที่ใช้พัฒนาโปรแกรมถูกจัดมาให้โดยไม่ต้องเสียค่าใช้จ่าย เทียบกับสมัยนั้นที่เครื่องมือพัฒนาโปรแกรมดี ๆ บนบางระบบมีราคาสูง

แน่นอนว่าถ้ามองเปรียบเทียบกับปัจจุบัน จุดเด่นเหล่านั้นไม่ได้เด่นอีกต่อไปแล้ว เพราะภาษาสมัยใหม่จำนวนมากก็มีคุณสมบัติเหล่านั้นเหมือนกัน แต่เมื่อเรามองย้อนไปเมื่อปี 1995 การที่ Java เปิดตัวมาพร้อมกับคุณสมบัติทั้งหมดนั้นทำให้ Java ได้รับความนิยมอย่างล้นหลาม และยังคงเป็นหนึ่งในภาษาที่ได้รับความนิยมที่สุดในปัจจุบันอยู่

ในช่วงปี 2009-2010 Oracle ได้ซื้อกิจการของ Sun Microsystems ไป และพัฒนา Java ต่อมาจนถึงปัจจุบัน

Oracle ได้จัดชุดพัฒนาโปรแกรมด้วยภาษา Java ไว้หลายรูปแบบ โดยรูปแบบหลัก ๆ มีดังนี้

- Java SE (Java Platform, Standard Edition) สำหรับการพัฒนาซอฟต์แวร์ทั่วไป
- Java EE (Java Platform, Enterprise Edition) เป็นส่วนขยายจาก Java SE โดยเพิ่มข้อกำหนดในส่วนของการพัฒนาระบบเว็บ การประมวลผลแบบกระจาย และเครื่องมืออื่น ๆ ที่ใช้ในการพัฒนาซอฟต์แวร์ระดับองค์กร
- Java ME (Java Platform, Micro Edition) เน้นที่การพัฒนาซอฟต์แวร์สำหรับอุปกรณ์ฝังตัวและอุปกรณ์เคลื่อนที่ เช่น ไมโครคอนโทรลเลอร์ โทรศัพท์มือถือ เป็นต้น ในทางปฏิบัติ ผู้พัฒนาฮาร์ดแวร์มักจะพัฒนาระบบสนับสนุนของตัวเองซึ่งมี API ที่แตกต่างกันไปขึ้นมาด้วย ประกอบกับการมาของระบบปฏิบัติการ Android พร้อมเครื่องมือพัฒนาของตัวเอง ทำให้ความนิยม Java ME ลดลงไปอย่างมาก

ภาษา Java ผ่านประวัติศาสตร์การใช้งานมาอย่างยาวนาน เทคโนโลยีเสริมที่พัฒนาจากภายนอกทำให้ความจำเป็นที่จะต้องใช้ Java EE และ Java ME ลดลงไปมาก เราสามารถใช้ Java SE ในการทำทุกอย่างที่ Java EE และ Java ME ทำได้โดยใช้ไลบรารีหรือเฟรมเวิร์กอื่น ๆ ช่วย

เราจะใช้ Java SE ในวิชานี้

## การติดตั้ง Java Development Kit

Java Development Kit (JDK) คือชุดพัฒนาโปรแกรมด้วยภาษา Java ซึ่งจะประกอบด้วย Java Virtual Machine (JVM) ที่ใช้ในการรันโปรแกรม ตัวไลบรารีต่าง ๆ และเครื่องมืออื่น ๆ ที่ช่วยในการพัฒนาโปรแกรม

ดาวน์โหลด JDK ได้ที่ http://www.oracle.com/technetwork/java/javase/downloads/index.html

เมื่อติดตั้งเสร็จแล้ว ลองทดสอบด้วยการพิมพ์คำสั่งต่อไปนี้ที่ Command Prompt (Windows), Terminal (Mac OSX) หรือ Shell (Linux)

`javac`

ถ้ารันได้ตามปกติแสดงว่าการติดตั้งเรียบร้อยดี ถ้าไม่พบ `javac` อาจจะเป็นเพราะต้องตั้งค่า `PATH` ให้เรียบร้อยก่อน บนระบบปฏิบัติการ Windows จะอยู่ที่ "C:\Program Files\Java\jdk1.x.x_xxx\bin" โดย x ในที่นี้คือรุ่นของ JDK ที่ติดตั้ง

นอกจาก JDK แล้วเรายังต้องมีเครื่องมือสำหรับเขียนโปรแกรมด้วย ในเบื้องต้นเราจะใช้ Atom ซึ่งดาวน์โหลดได้ที่ https://atom.io

## โปรแกรมแรกด้วยภาษา Java

ใช้ Atom เขียนโค้ดต่อไปนี้ แล้วบันทึกโดยตั้งชื่อไฟล์นี้ว่า Welcome.java

In [1]:
public class Welcome {
    public static void main(String[] args) {
        System.out.println("Welcome to the Java World!");
    }
}

### การคอมไพล์

การตั้งชื่อไฟล์โปรแกรมเราจะตั้งตามชื่อของคลาส (class) จากโค้ดด้านบนนี้ คลาสมีชื่อว่า `Welcome` เราจึงตั้งชื่อไฟล์ว่า Welcome.java

โปรแกรมที่เขียนด้วยภาษา Java ไม่สามารถรันได้ทันที ต้องผ่านขั้นตอนการคอมไพล์ก่อนเพื่อเปลี่ยนโค้ดภาษา Java ให้เป็นรหัสคำสั่งที่ JVM สามารถทำงานได้

เราคอมไพล์โปรแกรมด้วยคำสั่งต่อไปนี้

`javac Welcome.java`

`javac` เป็นชื่อของคอมไพเลอร์ภาษา Java ซึ่งเราใช้ในการคอมไพล์โค้ดโปรแกรมให้เป็นสิ่งที่เราเรียกว่าไบต์โค้ด (bytecode) ซึ่งเป็นชุดคำสั่งของ JVM

จากการคอมไพล์ข้างต้น เราจะได้ผลลัพธ์เป็นไฟล์ชื่อ Welcome.class

### การรัน

ไฟล์ที่มีนามสกุล .class จะอยู่ในรูปของไบต์โค้ดและสามารถรันได้ทันที

เรารันโปรแกรมนี้ด้วยคำสั่ง

`java Welcome`

ซึ่งจะได้ผลลัพธ์เป็นข้อความ

`Welcome to the Java World!`

สังเกตว่าการสั่งรันโปรแกรมด้วยคำสั่ง `java` เราไม่จำเป็นต้องระบุนามสกุลของไฟล์ (ในที่นี้คือ .class) เนื่องจากว่าสิ่งที่ตามหลังคำสั่ง `java` ไม่ใช่ชื่อไฟล์ แต่เป็นชื่อของคลาสที่ต้องการรัน

คำสั่ง `java` เป็นการโหลดไฟล์ .class เข้าไปสู่ JVM หรือ Java Virtual Machine เพื่อให้โปรแกรมเริ่มต้นทำงาน JVM จะทำตัวเป็นเหมือนเครื่องคอมพิวเตอร์เสมือนที่รันอยู่บนเครื่องคอมพิวเตอร์จริง ๆ อีกที แนวคิด Write Once, Run Anywhere เกิดขึ้นได้ด้วยหลักการนี้ เนื่องจากเครื่องคอมพิวเตอร์จริง ๆ แต่ละระบบก็จะมีความแตกต่างกัน โปรแกรมที่คอมไพล์สำหรับระบบหนึ่งจะไม่สามารถไปทำงานบนอีกระบบได้ แต่ Java แก้ปัญหานี้โดยการให้มี JVM ซึ่งเป็นซอฟต์แวร์จำลองระบบคอมพิวเตอร์เสมือน โปรแกรมภาษา Java จะถูกคอมไพล์มาเป็นรหัสคำสั่งของ JVM ซึ่งจะทำให้โปรแกรมนี้สามารถรันได้ทุกที่ทุกระบบที่มี JVM อยู่

### คำอธิบายโปรแกรม

ส่วนที่เป็นการทำงานจริงของโปรแกรมนี้คือ

```java
System.out.println("Welcome to the Java World!");
```

ซึ่งเป็นคำสั่งให้แสดงผลข้อความตามที่ระบุออกทางจอภาพแล้วขึ้นบรรทัดใหม่

นอกเหนือจากบรรทัดนี้ ส่วนอื่น ๆ เป็นส่วนที่จำเป็นต้องมีตามข้อกำหนดของโปรแกรมที่เขียนด้วยภาษา Java

ข้อแรกคือ โค้ดโปรแกรมจะต้องอยู่ในคลาส (class) ซึ่งเป็นหน่วยของโปรแกรมหน่วยหนึ่งที่ประกอบด้วยข้อมูลและโค้ดที่จัดการกับข้อมูลเหล่านั้น เรายังไม่ได้ให้ความสำคัญกับคลาสในตอนนี้

ข้อที่สองคือ โปรแกรมจะเริ่มทำงานที่เมทอด (method) ที่ชื่อ `main` ที่อยู่ภายในคลาส ถ้าเปรียบเทียบกับภาษาอื่น เมทอดก็คือฟังก์ชันที่อยู่ภายในคลาส เราสามารถจะมีเมทอดจำนวนเท่าใดก็ได้ในคลาสหนึ่ง ๆ แต่การทำงานของโปรแกรมจะเริ่มต้นที่เมทอด `main` เท่านั้น

เพราะฉะนั้น โปรแกรม Java ขั้นต่ำที่รันได้จะต้องมีคลาส และมีเมทอด `main` โดยคลาสจะใช้ชื่ออะไรก็ได้ตามหลักการตั้งชื่อ

ในเบื้องต้นขอให้ใช้โครงสร้างนี้ในการเขียนโปรแกรมไปก่อน เราจะกลับมาพูดถึงรายละเอียดของแต่ละส่วนอีกครั้งในบทที่เหมาะสม

```java
public class ClassName {
    public static void main(String[] args) {
        // Your code here
    }
}
```

`ClassName` ให้เราตั้งเป็นชื่อคลาสตามต้องการ และเขียนโค้ดอยู่ภายในเมทอด `main`

#### การเขียนคอมเมนต์

เราสามารถระบุส่วนที่เป็นคอมเมนต์หรือหมายเหตุในโปรแกรมได้ 3 รูปแบบดังนี้

```java
/* This is a block comment. Use this when you have long comments,
   usually spanning multiple lines. */

// And this is a single-line comment. It ends at the end of the line.

/**
 * This is a javadoc comment. We can use 'javadoc' tool to automatically
 * generate documentation.
 */
```

### ตัวแปรและชนิดข้อมูล

ตัวแปรในภาษา Java จะต้องถูกประกาศและกำหนดชนิดข้อมูลก่อนการใช้งาน เมื่อตัวแปรถูกกำหนดชนิดแล้วจะไม่สามารถเปลี่ยนแปลงได้

ชนิดข้อมูลในภาษา Java จะมี 2 ประเภท ได้แก่ชนิดพื้นฐาน (primitive type) และชนิดตัวอ้างอิง (reference type)

ชนิดข้อมูลพื้นฐานมีดังต่อไปนี้

| ชนิดข้อมูล | ลักษณะ    | ตัวอย่าง | ขนาด (บิต) | ช่วงของค่า |
|---------|----------|--------|----------:|----------|
| boolean | ตรรกะ    | false, true | 1 |  |
| char    | อักขระ    | 'a', 'A', ' ', '0' | 16 | \u0000 ... \uFFFF |
| byte    | จำนวนเต็ม | 0, 1, -1, -127, 50 | 8 | 2<sup>8</sup> -128 .. 127 |
| short   | จำนวนเต็ม | 0, 1, -1, 1000, 2_000 | 16 | -32,768 .. 32,767 |
| int     | จำนวนเต็ม | 0, 1, -1, 200_000 | 32 | -2,147,483,648 .. 2,147,483,647 |
| long    | จำนวนเต็ม | 0L, 1L, -1L, 2000L | 64 | -9,223,372,036,854,775,808 .. 9,223,372,036,854,775,807 |
| float   | ค่าทศนิยม  | -1.0f, 0.49f, 3E8f | 32 | ประมาณ ±1.4×10<sup>-38</sup> .. ±3.4×10<sup>38</sup>, นัยสำคัญ 7-8 หลัก |
| double  | ค่าทศนิยม  | -1.0, 0.49, 3E8 | 64 | ประมาณ ±4.9×10<sup>-324</sup> .. ±1.7×10<sup>308</sup>, นัยสำคัญ 15-16 หลัก |

ข้อมูลชนิดตัวอ้างอิงคือข้อมูลที่มีชนิดเป็นคลาส อินเทอร์เฟซ อาร์เรย์ หรือ `enum` ข้อมูลชนิด `String` ในภาษา Java มีชนิดเป็นตัวอ้างอิง เพราะ `String` เป็นคลาส

ตัวแปรที่ประกาศชนิดเป็นชนิดพื้นฐานจะเก็บข้อมูลตามชนิดนั้นไว้ที่ตัวแปรโดยตรง แต่ตัวแปรที่ประกาศชนิดเป็นตัวอ้างอิงจะไม่ได้เก็บตัวข้อมูลโดยตรง แต่จะเก็บตัวอ้างอิงไปยังข้อมูล เราจะพูดถึงตัวอ้างอิงอีกครั้งในบทเรื่องคลาสและอ็อบเจกต์

#### การระบุชนิดของค่าตัวเลข

จากตารางเราจะเห็นว่าค่าตัวเลขชนิด `long` จะมีการระบุ `L` ตามท้ายเพื่อบอกว่าเป็นตัวเลขชนิด `long` และชนิด `float` ก็จะมีการระบุ `f` ตามหลังเช่นกัน

ถ้าไม่ระบุ เลขจำนวนเต็มจะถือเป็นชนิด `int` และเลขทศนิยมจะถึอเป็นชนิด `double`

#### รูปแบบการเขียนจำนวนเต็ม

ค่าจำนวนเต็มสามารถเขียนได้ในรูปฐาน 10 และในรูปฐานอื่น ๆ ดังนี้

| ฐาน                  | รูปแบบ             | ตัวอย่าง |
|----------------------|-------------------|--------|
| ฐาน 10 (decimal)     | ไม่ขึ้นต้นด้วย 0       | 1234, -15, 1_500 |
| ฐาน 2 (binary)       | ขึ้นต้นด้วย 0b หรือ 0B | 0b101, 0B1110011 -0b10_1100 |
| ฐาน 8 (octal)        | ขึ้นต้นด้วย 0         | 01_234, -0723, 0755 |
| ฐาน 16 (hexadecimal) | ขึ้นต้นด้วย 0x หรือ 0X | 0xAB_CDEF, -0x7F, 0x12ABCF |

เราสามารถใช้เครื่องหมาย `_` (underscore) ในการแบ่งหลักของตัวเลขได้ การแบ่งหลักไม่มีข้อบังคับว่าจะต้องแบ่งเป็นชุดละกี่หลัก ควรใช้เพื่อเพิ่มความชัดเจนในการแสดงตัวเลขตามความเหมาะสม

#### รูปแบบการเขียนจำนวนทศนิยม

เราสามารถเขียนเลขทศนิยมตามรูปปกติ หรือรูปยกกำลังก็ได้ เช่น -25.24, 0.41, .25, 1.5E-7, -2.54E12 เป็นต้น รูปยกกำลังจะประกอบด้วยตัวคูณซึ่งก็คือตัวเลขก่อนตัว 'E' และกำลังของ 10 ซึ่งก็คือเลขหลังตัว 'E' เช่น 1.54E-7 ก็จะหมายถึง 1.5×10<sup>-7</sup>

#### ข้อมูลชนิดตัวเลขกับการล้นและการน้อยเกินเก็บ

การเก็บข้อมูลในคอมพิวเตอร์มีข้อจำกัดที่ขนาดของข้อมูล จากตารางชนิดข้อมูลพื้นฐานเราจะเห็นว่าตัวเลขแต่ละชนิดมีการระบุช่วงของค่าที่เป็นไปได้อยู่ด้วย เช่น `int` มีการระบุไว้เป็นช่วงตั้งแต่ -2,147,483,648 ถึง 2,147,483,647 และ `byte` มีค่าได้ตั้งแต่ -128 ถึง 127 เป็นต้น

เราลองศึกษาพฤติกรรมของตัวเลขเมื่อเราพยายามเพิ่มค่าไปมากกว่าค่าสูงสุดที่เป็นไปได้จากส่วนของโค้ดต่อไปนี้

In [8]:
int value = Integer.MAX_VALUE;
System.out.printf("Now: value = %,d%n", value);
value = value + 1;
System.out.printf("After increment: value = %,d%n", value);
value = value - 1;
System.out.printf("After decrement: value = %,d%n", value);

Now: value = 2,147,483,647
After increment: value = -2,147,483,648
After decrement: value = 2,147,483,647


> **หมายเหตุ** ส่วนของโค้ดด้านบนนั้นไม่สามารถคอมไพล์ได้โดยตรง ต้องเขียนอยู่ในโครงของคลาสและเมทอด `main` จึงจะนำไปคอมไพล์ได้ แต่เพื่อความกระชับ ในบางครั้งเราจะละโครงของคลาสและเมทอด `main` ในเอกสาร เวลาที่ทดสอบจริงต้องเขียนโครงเพิ่มเข้ามาด้วย

เรากำหนดให้ตัวแปร `value` มีค่าเป็น `Integer.MAX_VALUE` ซึ่งเป็นค่าสูงสุดที่ตัวแปรชนิด `int` จะเก็บได้ เราพบว่าเมื่อเพิ่มค่าจากค่าสูงสุดไปอีก 1 ค่า ผลลัพธ์ที่ได้คือ `value` จะมาอยู่ที่ค่าต่ำสุดที่เป็นไปได้ (-2,147,483,648) และเมื่อลดค่าจากค่าต่ำสุดไป 1 ค่า ค่าของ `value` ก็จะกลับมาอยู่ที่ค่าสูงสุดอีก

พฤติกรรมที่เราสังเกตได้นี้คือการล้น (overflow) ของค่า ตัวเลขจำนวนเต็มในคอมพิวเตอร์โดยทั่วไปแล้วจะจัดเก็บในรูปแบบที่เรียกว่าส่วนเติมเต็มของสอง (two's complement) ซึ่งในรูปแบบนี้ หากค่าตัวเลขมีค่าเป็นบวกสูงสุดแล้ว การเพิ่มค่าไปอีกจะทำให้ค่าไปอยู่ทางฝั่งเลขลบที่ต่ำที่สุด ในทางกลับกัน หากค่าตัวเลขมีค่าเป็นค่าลบต่ำที่สุดแล้ว การลดค่าไปอีกจะทำให้ค่าไปอยู่ฝั่งบวกที่สูงที่สุดแทน

ภาษา Java ไม่มีการตรวจสอบการล้นของจำนวนเต็มโดยอัตโนมัติ เป็นหน้าที่ของโปรแกรมเมอร์ที่จะต้องเลือกใช้ชนิดข้อมูลที่ใหญ่เพียงพอสำหรับข้อมูลที่จะจัดเก็บ และระมัดระวังในการใช้งานไม่ให้เกิดการล้นได้

เรามาดูในส่วนของเลขทศนิยมกันบ้าง

In [13]:
System.out.println("Floating-point overflow test:");
System.out.println("Max double: " + Double.MAX_VALUE);
System.out.println("Added with a large value: " + (Double.MAX_VALUE + 1_000_000.0));
System.out.println("Adding very large values: " + (1.5E308 + 1.5E308));

Floating-point overflow test:
Max double: 1.7976931348623157E308
Added with a large value: 1.7976931348623157E308
Adding very large values: Infinity


เราจะเห็นว่าค่าที่มีขนาดมากที่สุดของจำนวนทศนิยมชนิด `double` คือ 1.7976931348623157×10<sup>308</sup> เมื่อเราทดสอบโดยการบวกค่าขนาดใหญ่เพิ่มไปจากค่าที่มากที่สุด พบว่าเราจะได้ผลลัพธ์เป็นค่ามากที่สุดเหมือนเดิม แต่เมื่อเราเอาค่าที่มีขนาดใหญ่มาก ใกล้เคียงกับค่ามากที่สุดสองตัวมาบวกกัน เราจะได้ผลลัพธ์เป็น `Infinity`

การเก็บเลขทศนิยมในคอมพิวเตอร์เป็นการเก็บค่าประมาณ โดยมีจำนวนตัวเลขนัยสำคัญ (significant digit) ที่จำกัด เช่น ของชนิด `float` มีตัวเลขนัยสำคัญประมาณ 7-8 หลัก และ `double` ประมาณ 15-16 หลัก

เมื่อเลขเข้าใกล้ค่าสูงสุด ความละเอียดในหลักต่ำ ๆ ก็จะลดลง การกระทำกับเลขที่มีขนาดเล็กกว่ามาก ๆ อาจจะไม่ส่งผลใด ๆ เลยเนื่องจากการปัดค่าเข้าหาเลขที่สูงกว่าเหมือนเดิม แต่ถ้าเอาเลขที่มีขนาดใหญ่มาก ๆ ใกล้ค่าสูงสุดมาบวกกัน ผลที่ได้จะมีขนาดใหญ่มากเกินกว่าจะเก็บได้ และทำให้เกิดการล้นขึ้น ซึ่งในกรณีของ Java จะมีการตรวจสอบการล้นของเลขทศนิยม และกำหนดผลลัพธ์ให้เป็น `Infinity`

พฤติกรรมนี้เหมือนกันทั้งเลขบวกและเลขลบที่มีขนาดใหญ่มาก ๆ

ลองพิจารณาเลขทศนิยมที่มีขนาดเล็กมาก ๆ ดูบ้าง

In [20]:
System.out.println("Floating-point underflow test:");
System.out.println("Min double: " + Double.MIN_VALUE);
System.out.println("Subtracting a small value: " + (Double.MIN_VALUE - 1.0E-50));
System.out.println("Multiplying very small values: " + (4.0E-320 * 4.0E-320));

Floating-point underflow test:
Min double: 4.9E-324
Subtracting a small value: -1.0E-50
Multiplying very small values: 0.0



หลักนัยสำคัญ

เลขทศนิยมใน Java ทั้งชนิด `float` และ `double` มีรูปแบบการจัดเก็บและกฎเกณฑ์การจัดการกับค่าต่าง ๆ ตามมาตรฐาน [IEEE754](https://en.wikipedia.org/wiki/IEEE_754) ซึ่งสามารถศึกษาดูได้หากสนใจ

การเปรียบเทียบค่า floating-point

ใช้ BigDecimal ถ้าต้องการความแม่นยำสำหรับเลขฐาน 10

char

String, string concatenation

### ตัวดำเนินการและลำดับการทำงาน

### ธรรมเนียมการตั้งชื่อในภาษา Java

#### tabs -> spaces

#### line breaking

- where to break
- 2 tabs for continuation

#### เพิ่มเติม

1. https://google.github.io/styleguide/javaguide.html
2. https://en.wikipedia.org/wiki/Indentation_style

### การแสดงข้อความ

### การรับข้อมูล

In [2]:
System.out.println(1_234);

1234


No Outputs

In [11]:
import java.util.Scanner;

Scanner input = new Scanner(System.in);
String s = input.nextLine();
System.out.println(s);

EvalException: No line found

In [10]:
import java.io.BufferedReader;
import java.io.InputStreamReader;

BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
String name = reader.readLine();
System.out.println(name);

null
